### PR TITLE
Add support for NSX-T ALB Pools

### PR DESCRIPTION
**Pull request by Didainius**
This PR adds NSX-T ALB Pool support via `vcd_nsxt_alb_pool` resource and data source.

Testing is quite tricky because some prerequisites need to be created using System user and the ALB pool itself can be created by System or Org admin.  Due to limitations of Terraform acceptance test framework having multiple access levels there are two tests now:
* `TestAccVcdNsxtAlbPool` which can be run only using System user and builds up prerequisites using Terraform resources.
* `TestAccVcdNsxtAlbPoolOrgUser` which respects `VCD_TEST_ORG_USER=1` flag but uses SDK to setup prerequisites

This PR also changed a few references in `resource_vcd_nsxt_alb_edgegateway_service_engine_group_test.go` (e0240fc) for tests `TestAccVcdNsxtEdgeGatewayServiceEngineGroup*` because in some cases it could have caused race condition and errors.

Test with tag `alb` passed on 10.2 and 10.3 with `VCD_TEST_ORG_USER=1` and without it. (System user and Org user). Also binary tests and upgrade tests passed with tag `alb`

### CODE REVIEW

File: vcd/datasource_vcd_nsxt_alb_pool.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+					Schema: map[string]*schema.Schema{
+						"enabled": &schema.Schema{
+							Type:        schema.TypeBool,
+							Computed:    true,
+							Description: "Shows is the member is enabled or not",
+						},
+						"ip_address": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "Name of pool member",
```
Comment (vbauzysvmware) [inline vcd/datasource_vcd_nsxt_alb_pool.go:79]: Is this description is correct ?

File: vcd/datasource_vcd_nsxt_alb_pool.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+				},
+			},
+			"health_monitor": {
+				Type:     schema.TypeSet,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"type": &schema.Schema{
+							Type:        schema.TypeString,
+							Required:    true,
```
Comment (vbauzysvmware) [inline vcd/datasource_vcd_nsxt_alb_pool.go:119]:
```suggestion
							Computed:    true,
```
?

File: vcd/datasource_vcd_nsxt_alb_pool.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+			},
+			"ca_certificate_ids": {
+				Type:        schema.TypeSet,
+				Computed:    true,
+				Description: "A set of root certificate IDs to use when validating certificates presented by pool members",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"cn_check_enabled": &schema.Schema{
```
Comment (vbauzysvmware) [inline vcd/datasource_vcd_nsxt_alb_pool.go:164]: would be check with team if `cn_check_enabled` or `common_name_check_enabled`

File: vcd/resource_vcd_nsxt_alb_pool.go [hunk 4]
```diff
@@ -0,0 +1,646 @@
...
+			},
+			"description": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Description of ALB Pool",
+			},
+			"algorithm": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Algorithm for choosing pool members (default LEAST_CONNECTIONS)",
```
Comment (vbauzysvmware) [inline vcd/resource_vcd_nsxt_alb_pool.go:70]: Maybe is worth adding possible supported values?

File: website/docs/r/nsxt_alb_pool.html.markdown [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+  ca_certificate_ids = [data.vcd_library_certificate.sample-cert.id]
+  cn_check_enabled   = true
+  domain_names       = ["domain1", "domain2"]
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
```
Comment (vbauzysvmware) [inline website/docs/r/nsxt_alb_pool.html.markdown:95]: Org and VDC missing :)

**Review by vbauzysvmware — CHANGES_REQUESTED** [review]
First scan :)

File: vcd/datasource_vcd_nsxt_alb_pool.go (hunk 1 above)
Comment (Didainius) [reply vcd/datasource_vcd_nsxt_alb_pool.go]: As usual - some slip and I fixed.

File: vcd/datasource_vcd_nsxt_alb_pool.go (hunk 2 above)
Comment (Didainius) [reply vcd/datasource_vcd_nsxt_alb_pool.go]: As usual - some slip and I fixed.

File: vcd/datasource_vcd_nsxt_alb_pool.go (hunk 3 above)
Comment (Didainius) [reply vcd/datasource_vcd_nsxt_alb_pool.go:164]: cc @lvirbalas, @dataclouder

File: vcd/resource_vcd_nsxt_alb_pool.go (hunk 4 above)
Comment (Didainius) [reply vcd/resource_vcd_nsxt_alb_pool.go]: I didn't want to make it lengthy. But added now.

File: vcd/resource_vcd_nsxt_alb_pool.go [hunk 6]
```diff
@@ -0,0 +1,646 @@
...
+	if len(resourceURI) != 4 {
+		return nil, fmt.Errorf("resource name must be specified as org-name.vdc-name.nsxt-edge-gw-name.pool_name")
+	}
+	orgName, vdcName, edgeName, poolName := resourceURI[0], resourceURI[1], resourceURI[2], resourceURI[3]
+
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdc(orgName, vdcName)
+	if err != nil {
+		return nil, fmt.Errorf("unable to find Org %s: %s", vdcName, err)
```
[the remark this answers is not in the document]
Comment (Didainius) [reply vcd/resource_vcd_nsxt_alb_pool.go]: Changed

File: vcd/resource_vcd_nsxt_alb_pool.go [hunk 7]
```diff
@@ -0,0 +1,646 @@
...
+	albPools, err := vcdClient.GetAllAlbPools(edge.EdgeGateway.ID, nil)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve NSX-T ALB Pool '%s': %s", poolName, err)
+	}
+
+	if len(albPools) < 1 {
+		return nil, fmt.Errorf("ALB Pool with name '%s' in Edge Gateway '%s' not found", poolName, edgeName)
+	}
+
+	if len(albPools) > 1 {
```
[the remark this answers is not in the document]
Comment (Didainius) [reply vcd/resource_vcd_nsxt_alb_pool.go]: In fact no. And this code could have been simplified.  Changed it.

File: website/docs/r/nsxt_alb_pool.html.markdown (hunk 5 above)
Comment (Didainius) [reply website/docs/r/nsxt_alb_pool.html.markdown:95]: Indeed. I saved on space. Added.

File: vcd/datasource_vcd_nsxt_alb_pool.go (hunk 3 above)
Comment (lvirbalas) [reply vcd/datasource_vcd_nsxt_alb_pool.go:164]: This one is tricky. We need to look at all certificate related fields:
- `ca_certificate_ids`
- `cn_check_enabled`
- `domain_names`

In this light, "CN" is as common abbreviation as "CA" is, but I agree that `common_name_check_enabled` is clearer. It kind of becomes consistent with `domain_names` field then:
- `ca_certificate_ids`
- `common_name_check_enabled`
- `domain_names`

Alternatively, maybe there's a way to group all of them with some short prefix?

CC: need @dataclouder input too.

File: website/docs/r/nsxt_alb_pool.html.markdown [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+* `health_monitor` - (Optional) A block to define health monitor. Multiple can be used. See [Health
+  monitor](#health-monitor-block) and example for usage details.
+
+<a id="member-block"></a>
+## Member
+
+* `ip_address` (Required) IP address of pool member. 
+* `enabled` (Optional) defines if the pool member is enabled to receive traffic (default `true`)
+* `port` (Optional) Port for receiving traffic - overrides the root value `default_port` for individual members
+* `ratio` (Optional) Ratio of selecting eligible servers in the pool. (default `1`)
```
Comment (lvirbalas) [inline website/docs/r/nsxt_alb_pool.html.markdown:137]:
```suggestion
* `ratio` (Optional) Ratio of selecting eligible servers in the pool (default `1`)
```

File: website/docs/r/nsxt_alb_pool.html.markdown [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+  Cloud Director to switch the pool to a profile managed by Cloud Director. See [Persistence
+  profile](#persistence-profile-block) and example for usage details.
+* `health_monitor` - (Optional) A block to define health monitor. Multiple can be used. See [Health
+  monitor](#health-monitor-block) and example for usage details.
+
+<a id="member-block"></a>
+## Member
+
+* `ip_address` (Required) IP address of pool member. 
+* `enabled` (Optional) defines if the pool member is enabled to receive traffic (default `true`)
```
Comment (lvirbalas) [inline website/docs/r/nsxt_alb_pool.html.markdown:135]:
```suggestion
* `enabled` (Optional) Defines if the pool member is enabled to receive traffic (default `true`)
```

File: website/docs/r/nsxt_alb_pool.html.markdown [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+
+  * `CUSTOM_HTTP_HEADER` - Custom, static mappings of header values to specific servers are used. Header name must be provided as `value`
+
+  * `APP_COOKIE` - Load Balancer reads existing server cookies or URI embedded data such as JSessionID. Cookie name must be provided as `value`
+
+  * `TLS` - Information is embedded in the client's SSL/TLS ticket ID. This will use default system profile System-Persistence-TLS
+
+* `value` (Optional) is required for some `type`s: ``HTTP_COOKIE`, `CUSTOM_HTTP_HEADER`, `APP_COOKIE`
+
+### Attributes of persistence profile
```
Comment (lvirbalas) [inline website/docs/r/nsxt_alb_pool.html.markdown:162]: I think the quoting has been lost at this point, as all the lines starting this one show up as blue. Looks like we have "a quote issue" in the above line, Houston :)

**Review by lvirbalas — CHANGES_REQUESTED** [review]
Doc comments.

File: website/docs/r/nsxt_alb_pool.html.markdown (hunk 8 above)
Comment (Didainius) [reply website/docs/r/nsxt_alb_pool.html.markdown]: Fixed.

File: website/docs/r/nsxt_alb_pool.html.markdown (hunk 9 above)
Comment (Didainius) [reply website/docs/r/nsxt_alb_pool.html.markdown]: Fixed.

File: website/docs/r/nsxt_alb_pool.html.markdown (hunk 10 above)
Comment (Didainius) [reply website/docs/r/nsxt_alb_pool.html.markdown:162]: Yes. Indeed it was. I think I fixed it and improved the message just a little bit.

File: vcd/datasource_vcd_nsxt_alb_pool.go [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+						},
+						"ratio": {
+							Type:        schema.TypeInt,
+							Computed:    true,
+							Description: "Load ratio",
+						},
+						"marked_down_by": {
+							Type:        schema.TypeSet,
+							Computed:    true,
+							Description: "Marked down by",
```
Comment (lvirbalas) [inline vcd/datasource_vcd_nsxt_alb_pool.go:94]: Better add more in-depth description from the docs.

File: vcd/datasource_vcd_nsxt_alb_pool.go (hunk 3 above)
Comment (lvirbalas) [reply vcd/datasource_vcd_nsxt_alb_pool.go:164]: This morning the `cn_check_enabled` look fine to me :)

File: vcd/resource_vcd_nsxt_alb_pool.go [hunk 12]
```diff
@@ -0,0 +1,639 @@
...
+
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdc(orgName, vdcName)
+	if err != nil {
+		return nil, fmt.Errorf("unable to find VDC %s: %s", vdcName, err)
+	}
+
+	if vdc.IsNsxv() {
+		return nil, fmt.Errorf("ALB Pools are only supported on NSX-T please use 'vcd_lb_server_pool' for NSX-V load balancers")
```
Comment (lvirbalas) [inline vcd/resource_vcd_nsxt_alb_pool.go:361]: If you're capitalizing "Pools" then probably "Load Balancers" should be as well?
```suggestion
		return nil, fmt.Errorf("ALB Pools are only supported on NSX-T please use 'vcd_lb_server_pool' for NSX-V Load Balancers")
```

File: vcd/datasource_vcd_nsxt_alb_pool.go (hunk 11 above)
Comment (Didainius) [reply vcd/datasource_vcd_nsxt_alb_pool.go]: Changed in both resource and data source to 
```go
Description: "Marked down by provides a set of health monitors that marked the service down",
```

File: vcd/resource_vcd_nsxt_alb_pool.go (hunk 12 above)
Comment (Didainius) [reply vcd/resource_vcd_nsxt_alb_pool.go]: Changed and this and a few more capitalization cases

**Review by lvirbalas — APPROVED** [review]

File: vcd/datasource_vcd_nsxt_alb_pool.go (hunk 3 above)
Comment (dataclouder) [reply vcd/datasource_vcd_nsxt_alb_pool.go:164]: I prefer `common_name_check_enabled`

**Review by dataclouder — APPROVED** [review]

**Review by vbauzysvmware — APPROVED** [review]

**Pull request merged**